src/urbit/app/chess.hoon

```diff
@@ -658,12 +810,47 @@
               ?~  new.move-result
                 [cards.move-result this]  ::  nice try, cheater
               =,  u.new.move-result
-              :-  cards.move-result
-              ?.  ?=(~ result.game)
+              ::
+              ::  XX: shouldn't notify if %chess is open
+              ::  XX: post-Groups 2, should create more granular
+              ::      notifications about moves:
+              ::      '~sampel-palnet moves E5 pawn to E6'
+              ::      '~sampel-palnet: check!'
+              ::      '~sampel-palnet: checkmate!'
+              ::      '~sampel-palnet promotes XX pawn to queen'
+              ::
+              ::  check if opponent has resigned
+              ::  XX: must change %end check after GitHub issue #3 implemented
+              ::  XX: notification: opponent resigned
+              ?:  ?=(%end -.move)
+                :-  cards.move-result
                 %=  this
                   games    (~(del by games) u.game-id)
                   archive  (~(put by archive) u.game-id game)
                 ==
+              :-
+              %+  welp
+                cards.move-result
+              :~  :*  %pass   /hark-store
+                      %agent  [our.bowl %hark-store]
+                      %poke   :-  %hark-action
+                              !>  :-  %unread-count
+                                  :-  %chess
+                                  :-  /games/(scot %da u.game-id)/moves
+                                  :-  %.y  1
+                  ==
+                  :*  =/  title=(list content:hark)
+                        ~[[%text (crip "{<src.bowl>} has made a move")]]
+                      =/  =bin:hark
+                        [/chess/games [%chess /games/(scot %da u.game-id)/updates]]
+                      =/  hark-action=action:hark
+                        ::  XX: should link to game
+                        [%add-note bin title ~ now.bowl / /chess]
+                      =/  =cage
+                        [%hark-action !>(hark-action)]
+                      [%pass /hark-store %agent [our.bowl %hark-store] %poke cage]
+                  ==
+              ==
```

Comment:
It's not difficult to add more granular move notifications right now by moving the generation of this card into `++try-move` instead. I wrote most of the code for it while playing around:
```
++  try-move
  |=  [game=chess-game position=chess-position move=chess-move]
  ^-  [new=(unit [game=chess-game position=chess-position]) cards=(list card)]
  ?.  ?=(~ result.game)
    [~ ~]
  =/  new-position
    (~(apply-move with-position position) move)
  ?~  new-position
    [~ ~]
  =/  updated-game  `chess-game`game
  =.  moves.updated-game  (snoc moves.updated-game move)
  =/  fen  (position-to-fen u.new-position)
  =/  in-checkmate  ~(in-checkmate with-position u.new-position)
  =/  in-stalemate  ?:  in-checkmate
                      |
                    ~(in-stalemate with-position u.new-position)
  =/  base-cards
    =/  position-update-card
      :*  %give  %fact  ~[/game/(scot %da game-id.game)/updates]
          %chess-update  !>([%position game-id.game fen])
      ==
    ?:  xx  ::  XX our turn?
      [position-update-card ~]
    =/  place
      :-  %chess
      ?:  |(in-checkmate in-stalemate)
        /games
      /games/(scot %da u.game-id)
    =/  notification
      %+  rap  3
      :-  xx  ::  XX opponent name
      ?:  in-checkmate
        ' wins. Checkmate!'
      ?:  in-stalemate
        ' forced a stalemate'
      :-  (~(algebraicize with-position u.new-position) move)
      ?.  ~(in-check with-position u.new-position)
        ''
      '. Check!'
    :~  position-update-card
        :*  =/  title=(list content:hark)
              [[%text notification] ~]
            =/  =bin:hark
              [/chess/games [%chess place]]
            =/  hark-action=action:hark
              [%add-note bin title ~ now.bowl / /chess]
            =/  =cage
              [%hark-action !>(hark-action)]
            [%pass /hark-store %agent [our.bowl %hark-store] %poke cage]
        ==
    ==
  =/  result-cards
    ?:  ?|  ?=(%end -.move)
            in-checkmate
            in-stalemate
        ==
      =.  result.updated-game
        ?:  ?=(%end -.move)  `+.move
        ?:  in-stalemate  `%'½–½'
        ?:  in-checkmate
          ?-  player-to-move.u.new-position
            %white  `%'0-1'
            %black  `%'1-0'
          ==
        !!
      :*  :*  %give  %fact  ~[/game/(scot %da game-id.game)/updates]
              %chess-update
              !>([%result game-id.game (need result.updated-game)])
          ==
          :*  %give  %kick  :~  /game/(scot %da game-id.game)/updates
                                /game/(scot %da game-id.game)/moves
                            ==
              ~
          ==
          base-cards
      ==
    base-cards
  [`[updated-game u.new-position] result-cards]
--
```

In addition: should we be adding a `%del-place` card once a game ends, to delete all past/future notifications for a game that's over?

Reply:
I thought about `%del-place` but the only issue is: if `~sampel-palnet` wins or ends the game while I'm away, and we send a Grid notification for that, how do we avoid that notification being deleted as well? Cards are "processed" at exactly the same time so we can't just arrange them in the order we'd like them to be actioned.

Reply:
Current state of the `++try-move` helper core, now a door so it can access `bowl`. (Otherwise unchanged. The helper core is correctly positioned as-is.) The "~zod has made a move" notification works but the notifications defined in `++try-move` don't.

```
|_  =bowl:gall
++  try-move
  |=  [game=chess-game position=chess-position move=chess-move]
  ^-  [new=(unit [game=chess-game position=chess-position]) cards=(list card)]
  ?.  ?=(~ result.game)
    [~ ~]
  =/  new-position
    (~(apply-move with-position position) move)
  ?~  new-position
    [~ ~]
  =/  updated-game  `chess-game`game
  =.  moves.updated-game  (snoc moves.updated-game move)
  =/  fen  (position-to-fen u.new-position)
  =/  in-checkmate  ~(in-checkmate with-position u.new-position)
  =/  in-stalemate  ?:  in-checkmate
                      |
                    ~(in-stalemate with-position u.new-position)
  =/  ship-to-move
    ?-  player-to-move.position
      %white
        white.game
      %black
        black.game
    ==
  =/  base-cards
    =/  position-update-card
      :*  %give  %fact  ~[/game/(scot %da game-id.game)/updates]
          %chess-update  !>([%position game-id.game fen])
      ==
    ::  XX: why does this check [%ship @p] and not chess-player?
    ::      why include [%name @t] in chess-player at all if it breaks?
    ?>  ?=([%ship @p] ship-to-move)
    ?.  =(team:title +.ship-to-move)
      [position-update-card ~]
    =/  place
      :-  %chess
      ?:  |(in-checkmate in-stalemate)
        /games
      /games/(scot %da game-id.game)
    =/  notification
      %+  rap  3
        :~  `@`+.ship-to-move
            ?:  in-checkmate
              `@`': checkmate!'
            ?:  in-stalemate
              `@`' forced a stalemate'
            %+  rap  3
              :~
                `@`(~(algebraicize with-position u.new-position) move)
              ?.  ~(in-check with-position u.new-position)
                `@`''
              `@`', check!'
             ==
        ==
    :~  position-update-card
        :*  =/  title=(list content)
              [[%text notification] ~]
            =/  =bin
              [/chess/games [%chess place]]
            =/  hark-action=action:hark
              [%add-note bin title ~ now.bowl / /chess]
            =/  =cage
              [%hark-action !>(hark-action)]
            [%pass /hark-store %agent [our.bowl %hark-store] %poke cage]
        ==
    ==
  =/  result-cards
    ?:  ?|  ?=(%end -.move)
            in-checkmate
            in-stalemate
        ==
      =.  result.updated-game
        ?:  ?=(%end -.move)  `+.move
        ?:  in-stalemate  `%'½–½'
        ?:  in-checkmate
          ?-  player-to-move.u.new-position
            %white  `%'0-1'
            %black  `%'1-0'
          ==
        !!
      :*  :*  %give  %fact  ~[/game/(scot %da game-id.game)/updates]
              %chess-update
              !>([%result game-id.game (need result.updated-game)])
          ==
          :*  %give  %kick  :~  /game/(scot %da game-id.game)/updates
                                /game/(scot %da game-id.game)/moves
                            ==
              ~
          ==
          base-cards
      ==
    base-cards
  [`[updated-game u.new-position] result-cards]
--
```

Reply:
See the latest commit for current status. At the moment it outright crashes with a need-have error highlighted in a comment, but somehow I feel like that's progress from "looks correct but notification doesn't fire".

I got rid of `base-cards` while I was trying to refactor the core a little to make sense of it, as I kept getting lost in the nested `tisfas`s to keep track of the actual logic being executed. I feel like this core is almost working... but I wonder if it should be broken up into arms with all the conditional logic that matters clustered at the top.